validator/client/src/main/java/tech/pegasys/teku/validator/client/EpochDuties.java

```diff
@@ -56,21 +57,42 @@ public void onAttestationAggregationDue(final UInt64 slot) {
   }
 
   public int countDuties() {
-    return duties.map(ScheduledDuties::countDuties).orElse(0);
+    return getCurrentDuties().map(ScheduledDuties::countDuties).orElse(0);
+  }
+
+  public synchronized void recalculate() {
+    duties.cancel(false);
+    // We need to ensure the duties future is completed before .
+    duties = dutyLoader.loadDutiesForEpoch(epoch);
+    duties.finish(
+        this::processPendingActions,
+        error -> {
+          if (!(Throwables.getRootCause(error) instanceof CancellationException)) {
+            LOG.error("Failed to load duties", error);
+          } else {
+            LOG.trace("Loading duties cancelled", error);
+          }
+        });
   }
 
   public synchronized void cancel() {
-    futureDuties.cancel(false);
+    duties.cancel(false);
     pendingActions.clear();
   }
 
-  private synchronized void onDutiesLoaded(final ScheduledDuties scheduledDuties) {
-    duties = Optional.of(scheduledDuties);
+  private synchronized void processPendingActions(final ScheduledDuties scheduledDuties) {
     pendingActions.forEach(action -> action.accept(scheduledDuties));
     pendingActions.clear();
   }
 
   private synchronized void execute(final Consumer<ScheduledDuties> action) {
-    this.duties.ifPresentOrElse(action, () -> pendingActions.add(action));
+    getCurrentDuties().ifPresentOrElse(action, () -> pendingActions.add(action));
+  }
+
+  private synchronized Optional<ScheduledDuties> getCurrentDuties() {
+    if (!duties.isCompletedNormally()) {
+      return Optional.empty();
+    }
+    return Optional.of(duties.join());
```

Comment:
Does `.join()` here mean that we block this thread until current duties are completed? Is that okay?

Reply:
We already checked that the future is completed so we know it will return immediately.